devtools/gha/unittest.py

```diff
@@ -0,0 +1,52 @@
+import importlib
+import inspect
+import json
+import os
+from pathlib import Path
+import sys
+import unittest
+
+source = importlib.util.find_spec('nutils').origin
+assert source.endswith(os.sep + '__init__.py')
+source = source[:-11]
+coverage = {}
+
+if hasattr(sys, 'monitoring'):
+
+    def start(code, _):
+        if isinstance(code.co_filename, str) and code.co_filename.startswith(source) and not sys.monitoring.get_local_events(sys.monitoring.COVERAGE_ID, code):
+            if (file_coverage := coverage.get(code.co_filename)) is None:
+                with open(code.co_filename, 'rb') as f:
+                    nlines = sum(1 for _ in f)
+                coverage[code.co_filename] = file_coverage = [0] * (nlines + 1)
+            for _, _, l in code.co_lines():
+                if l:
+                    file_coverage[l] = 1
```

Comment:
Does this not imply that functions that are never visited do not count as misses?

Reply:
No, because we call `start` recursively for all constants in the code object, which includes functions defined in the code object.

Reply:
Neat.